### DISCUSSION TITLE
Child Component throws a 404 when the Model is deleted

### DISCUSSION THREAD

**Discussion by BVFI** [livewire/livewire#8564] [1. Help] [answered]
# The issue
Working with child components has been a nightmare for me. No error logs, difficult debugging, just guessing and here's my exact issue: I have a Listing and an Image model. The image model is mostly used as a child component, it handles all image uploads, savings and deletion of images. As I first started developing the component, I had in mind that the images would always be required in the listing, but now I have come to switch them to be optional and I missed something. Problem is, the child component doesn't leave a stack trace, an error log or anything, just a simple 404 page and "good luck, mate". Debugbar is a must when working with livewire. Heres the only clues I have to work with, but I cant pinpoint where exactly this is triggered:
```
- No query results for model [App\Models\Image].
- vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php#620
(Illuminate\Database\Eloquent\ModelNotFoundException):

        return $model;
    }
    throw (new ModelNotFoundException)->setModel(get_class($this->model));
}
```
In the Network tab there's also a:
```
Request URL: /listings/edit/null
Request Method: GET
Status Code: 404 Not Found
```
# The Code
The route:
```php
Route::middleware([
    'auth:sanctum',
    config('jetstream.auth_session'),
    'verified',
])->group(function () {
    Route::get('listings/edit/{listing}', Listing\Edit::class)->name('listings.edit');
    ...
```
The Listing edit.blade.php is:
```html
<livewire:image.image-upload 
    wire:key="{{ $uniqueKey }}" 
    :labelText="__('messages.photo_recommend', ['imagetype' => 'Header', 'format' => '960px:320px', 'min' => '200'])" 
    :imageSize="'w-full h-52'" :image="$header_image" wire:model="image">
</livewire:image.image-upload>
@if($og_image)
    <img class="w-full h-96" src="{{ $og_image }}" alt="Open Graph Image">
    ...
```
The image component(Dumbed down to focus on the image removal):
```php
class ImageUpload extends Component
{
    use WithFileUploads;
    #[Validate]
    public $image;
    #[Validate]
    public $alt;
    public $type;

    public $errorMsg = null, $imageSize = null, $labelText;

    public function mount(): void
    {
        if(($this->image instanceof Image)) {
            $this->fill(
                $this->image->only('alt')
            );
        }
    }
	
    public function updatedImage($value): void
    {
        \Debugbar::addMessage('updatedImage()', 'Image');
        if(!is_null($this->image)) {
            if (!$this->_validateImage()) $this->removeImage();
            $imagePath = $value->temporaryUrl();
            $this->dispatch('on-image-change', imagePath: $imagePath);
        }
    }
	
    public function removeImage(bool $completeRemove = false): void
    {
        \Debugbar::addMessage('removeImage()', 'Image');
        if($completeRemove && ($this->image instanceof Image)) {
            $path = 'public' .$this->image->path .$this->image->filename;
            \Debugbar::addMessage($path, 'Image Path');

            if ($this->image->delete()) {
                \Debugbar::addMessage('Image deleted from database', 'Image');
                //$this->image = new Image();
                \Debugbar::addMessage($this->image, 'Removed image');
                if (Storage::exists($path)) {
                    Storage::delete($path);
                    \Debugbar::addMessage('Image deleted from storage', 'Image');
                }
                \Debugbar::addMessage('Dispatching remove-image event', 'Image');
                $this->dispatch('remove-image'); // Goes in the Listing\Edit
            }
        }
        $this->reset();
//        $this->reset('image');
//        $this->reset('alt');

        //$this->dispatch('on-image-change', imagePath: '');
    }
}
```
The Listing\Edit component:
```php
class Edit extends Component
{
    public $header_image = null, $isImageValidated = true, $uniqueKey;
    public $meta_title, $meta_description, $og_title, $og_description, $og_image = '';
    public function mount()
    {
        $this->authorize('update', $this->listing);
        $this->uniqueKey = Random::generateRandomUniqueKey();
        $listingMeta = $this->listing->meta;
//      Pre-Fills fields
        $headerImageModel = $this->listing->images()->where('type', 'header')->first();
        if ($headerImageModel) {
            $this->header_image = $headerImageModel;
            $this->og_image = asset('/storage/' . $headerImageModel->path . $headerImageModel->filename);
        } else {
            $this->header_image = null;
        }
//      Pre-Fills inputs
        $this->fill(
            $this->listing->only('title', 'description', 'contact_name', 'contact_email', 'is_published'),
        );
    }

    public function render()
    {
        $this->sub_categories = Category::where('parent_id', $this->main_category)->get();
//        $parentId = Category::where('id', $this->listing->category_id)->pluck('parent_id');
        return view('livewire.listing.edit', [
            ...
            'main_categories',
            'sub_categories' => $this->sub_categories,
            'header_image' => $this->header_image,
        ]);
    }
    #[On('remove-image')]
    public function headerImageRemoved()
    {
        \Debugbar::addMessage('headerImageRemoved()', 'Listing');
        if ($this->listing && $this->header_image) {
            $this->header_image = null;
            $this->og_image = null;
        }
    }
	
    #[On('on-image-change')]
    public function updateOgImage($imagePath)
    {
        \Debugbar::addMessage('updateOgImage()', 'Listing');
        $this->og_image = $imagePath;
    }
	
	#[On('test-call')]
    public function test()
    {
        \Debugbar::addMessage('Very nice test', 'Listing');
    }
}
```
upload.blade.php
```blade
<div x-data="{imageName: null, imagePreview: {{ $image instanceof \App\Models\Image ? '\'' . asset('/storage/' . $image->path . $image->filename) . '\'' : 'null' }}}">
<x-input wire:model.blur="alt" id="alt" placeholder="Bildbeschreibung"></x-input>
<x-input-error for="alt" />
<div x-show="imagePreview" style="display: none;" x-on:remove-image.window="imagePreview = null">
    <span x-bind:style="'background-image: url(\'' + imagePreview + '\');'">
    </span>
</div>

{{--            Image Input--}}
<input type="file" id="image" wire:model="image" class="hidden" x-ref="image" x-on:change="handleImageChange" accept="image/png, image/jpeg, image/webp">
<x-secondary-button class="mt-2 me-2" type="button" x-on:click.prevent="$refs.image.click();">
    {{ __('messages.photo_new') }}
</x-secondary-button>
<x-danger-button wire:confirm="{{ __('messages.photo_delete') }}?" x-show="imagePreview"type="button" wire:click.prevent="removeImage(true)" {{--x-on:click="imagePreview = null"--}}>
{{ __('messages.photo_delete') }}
</x-danger-button>
    ...
<script>
    function handleImageChange(event) {
        const input = event.target;
        const file = input.files[0];
        const type = file.type;
        const maxSize = 8*1024*1024
        if(!['image/jpeg', 'image/png', 'image/webp'].includes(type)) {
            alert('Nur JPG, JPEG, PNG, und WEBP erlaubt.');
            return;
        }

        if(file.size > maxSize) {
            alert('Datei ist zu groß.');
            return;
        }

        if(type.startsWith('image')){
            const reader = new FileReader();
            reader.onload = (e) => {
                this.imageName = input.files[0].name;
                this.imagePreview = e.target.result;
            };
            reader.readAsDataURL(input.files[0]);
        }
    }
</script>
```
## Conclusion
Here's the interesting part: In the Image component, in the ```removeImage()``` function, If I make the code like this:
```php
public function removeImage(bool $completeRemove = false): void
    {
        if($completeRemove && ($this->image instanceof Image)) {
            $path = 'public' .$this->image->path .$this->image->filename;
            if ($this->image->delete())  $this->image = new Image();
                if (Storage::exists($path)) Storage::delete($path);
                $this->dispatch('test-call'); // Goes in the Listing\Edit
            }
        }
    }
```
If I just make it to delete from DB and storage and then call the ```test-call```, it still throws that 404 error.  If I comment out the ```dispatch('test-call')``` and no more events are called, then the image is actually removed, but the preview is still here. I have put a Debugbar at the beginning of each function and the last one I get is the ones in ```removeImage()```. When it goes in the ```test-call```, nothing is printed, nor is there a message. For some reason, after I delete the model, its still looking for it, I believe inside the ImageUpload component, but how can I tell?

So far, the only way to differentiate if the error comes from the parent or child component, has been if the error 404 was thrown inside of a new modal-window or on the same window. If its on the same window, the problem should be with the parent controller, if not the problem should be with the child component. 

It probably has something to do with Hydration of data, but it doesnt say where or what. My theory is that $this->image is re-hydrated and somewhere it breaks the code, because its not handled, but **where**?

# Things I have tried
## Read and hopefully didnt miss something in the [Nesting components documentation](https://livewire.laravel.com/docs/nesting)
> Another powerful pattern for sharing state between parent and child components is using wire:model directly on a child component via Livewire's Modelable feature.

- Already have a ```wire:model="image"``` in my parent controller.

At [this part](https://livewire.laravel.com/docs/nesting#listening-for-events-from-children) there's a sample code and it has the basic functionality as mine, so this leads me to believe there is something else wrong.
[Forcing a child component to render](https://livewire.laravel.com/docs/nesting#forcing-a-child-component-to-re-render) did nothing, regardless if I use ```wire:key``` or ```:key```

## [404 on Deletion with Child Component](https://github.com/livewire/livewire/discussions/5032)
- Suggest that wire:key was the issue. I already had a wire:key to identify each child component instance. I also can't make it delete the model from the parent (Unless I use events, I guess), because my image upload child component should handle that, otherwise whats the purpose of child components If they have their functionality split up between parents. I dont want to decentralize my app so much, that I have to play hide'n'seek with what everything does.  I need the child component, which handles one Model, to be able to delete that model. 
## [[v3.0.3] 404 Error after deleting item via parent component event](https://github.com/livewire/livewire/discussions/6792)
- Computed doesn't help, because it already knows the model is null.
Im tired, boss, Ive been debugging this for 2 days, trying different variations and changing stuff, to no avail. I don't know anymore.

**Comment by BVFI** [comment, marked as answer]
I have finally made some progress, after removing most of the code so that I only have the issue focused. Seems like making the ```$header_image``` public was the issue, so I transfered the ```$header image``` in the ```render()``` method,  and then I only work with the ```$this->image``` from the ```ImageUpload``` component and send it via events whenever its needed in the parent controller. 
This now means I have to rework most of my Listing functionality, which is just great, what a great day...

**Reply by Joystickiboy** [reply]
Man thanks for replying to yourself with the answer. I've been on a very similar issue for the past 2 days and just didn't understand what could possible cause this issue. You're a real king for this. Thanks
